File: client/cpp/hardware/hardware.cpp
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include "client/cpp/errors/errors.h"
#include "client/cpp/hardware/hardware.h"
#include "x/cpp/xerrors/errors.h"

namespace synnax {
Rack::Rack(const RackKey key, std::string name): key(key), name(std::move(name)) {}

Rack::Rack(std::string name): name(std::move(name)) {}

Rack::Rack(const api::v1::Rack &rack): key(rack.key()), name(rack.name()) {}

void Rack::to_proto(api::v1::Rack *rack) const {
    rack->set_key(key);
    rack->set_name(name);
}

const std::string RETRIEVE_RACK_ENDPOINT = "/hardware/rack/retrieve";
const std::string CREATE_RACK_ENDPOINT = "/hardware/rack/create";

std::pair<Rack, xerrors::Error> HardwareClient::retrieve_rack(const RackKey key) const {
    auto req = api::v1::HardwareRetrieveRackRequest();
    req.add_keys(key);
    auto [res, err] = rack_retrieve_client->send(RETRIEVE_RACK_ENDPOINT, req);
    if (err) return {Rack(), err};
    if (res.racks_size() == 0)
        return {
            Rack(),
            xerrors::Error(
                xerrors::NOT_FOUND,
                "Rack matching" + std::to_string(key) + " not found"
            )
        };
    auto rack = Rack(res.racks(0));
    rack.tasks = TaskClient(
        rack.key,
        task_create_client,
        task_retrieve_client,
        task_delete_client
    );
    return {rack, err};
}

std::pair<Rack, xerrors::Error>
HardwareClient::retrieve_rack(const std::string &name) const {
    auto req = api::v1::HardwareRetrieveRackRequest();
    req.add_names(name);
    auto [res, err] = rack_retrieve_client->send(RETRIEVE_RACK_ENDPOINT, req);
    if (err) return {Rack(), err};
    if (res.racks_size() == 0)
        return {
            Rack(),
            xerrors::Error(xerrors::NOT_FOUND, "Rack matching" + name + " not found")
        };
    if (res.racks_size() > 1)
        return {
            Rack(),
            xerrors::Error(
                xerrors::MULTIPLE_RESULTS,
                "Multiple racks matching" + name + " found"
            )
        };
    auto rack = Rack(res.racks(0));
    rack.tasks = TaskClient(
        rack.key,
        task_create_client,
        task_retrieve_client,
        task_delete_client
    );
    return {rack, err};
}

xerrors::Error HardwareClient::create_rack(Rack &rack) const {
    auto req = api::v1::HardwareCreateRackRequest();
    rack.to_proto(req.add_racks());
    auto [res, err] = rack_create_client->send(CREATE_RACK_ENDPOINT, req);
    if (err) return err;
    if (res.racks_size() == 0) return unexpected_missing("rack");
    rack.key = res.racks().at(0).key();
    rack.tasks = TaskClient(
        rack.key,
        task_create_client,
        task_retrieve_client,
        task_delete_client
    );
    return err;
}

std::pair<Rack, xerrors::Error>
HardwareClient::create_rack(const std::string &name) const {
    auto rack = Rack(name);
    auto err = create_rack(rack);
    return {rack, err};
}

xerrors::Error HardwareClient::delete_rack(const RackKey key) const {
    auto req = api::v1::HardwareDeleteRackRequest();
    req.add_keys(key);
    auto [res, err] = rack_delete_client->send(CREATE_RACK_ENDPOINT, req);
    return err;
}

Task::Task(
    TaskKey key,
    std::string name,
    std::string type,
    std::string config,
    bool internal,
    bool snapshot
):
    key(key),
    name(std::move(name)),
    type(std::move(type)),
    config(std::move(config)),
    internal(internal),
    snapshot(snapshot) {}

Task::Task(
    std::string name,
    std::string type,
    std::string config,
    bool internal,
    bool snapshot
):
    key(create_task_key(0, 0)),
    name(std::move(name)),
    type(std::move(type)),
    config(std::move(config)),
    internal(internal),
    snapshot(snapshot) {}

Task::Task(
    RackKey rack,
    std::string name,
    std::string type,
    std::string config,
    bool internal,
    bool snapshot
):
    key(create_task_key(rack, 0)),
    name(std::move(name)),
    type(std::move(type)),
    config(std::move(config)),
    internal(internal),
    snapshot(snapshot) {}

Task::Task(const api::v1::Task &task):
    key(task.key()),
    name(task.name()),
    type(task.type()),
    config(task.config()),
    internal(task.internal()),
    snapshot(task.snapshot()) {}

void Task::to_proto(api::v1::Task *task) const {
    task->set_key(key);
    task->set_name(name);
    task->set_type(type);
    task->set_config(config);
    task->set_internal(internal);
    task->set_snapshot(snapshot);
}

const std::string RETRIEVE_TASK_ENDPOINT = "/hardware/task/retrieve";
const std::string CREATE_TASK_ENDPOINT = "/hardware/task/create";
const std::string DELETE_TASK_ENDPOINT = "/hardware/task/delete";

std::pair<Task, xerrors::Error> TaskClient::retrieve(const TaskKey key) const {
    auto req = api::v1::HardwareRetrieveTaskRequest();
    req.set_rack(rack);
    req.add_keys(key);
    auto [res, err] = task_retrieve_client->send(RETRIEVE_TASK_ENDPOINT, req);
    if (err) return {Task(), err};
    if (res.tasks_size() == 0)
        return {
            Task(),
            xerrors::Error(
                xerrors::NOT_FOUND,
                "Task matching" + std::to_string(key) + " not found"
            )
        };
    return {Task(res.tasks(0)), err};
}

std::pair<Task, xerrors::Error> TaskClient::retrieve(const std::string &name) const {
    auto req = api::v1::HardwareRetrieveTaskRequest();
    req.set_rack(rack);
    req.add_names(name);
    auto [res, err] = task_retrieve_client->send(RETRIEVE_TASK_ENDPOINT, req);
    if (err) return {Task(), err};
    if (res.tasks_size() == 0)
        return {
            Task(),
            xerrors::Error(xerrors::NOT_FOUND, "Task matching" + name + " not found")
        };
    return {Task(res.tasks(0)), err};
}

std::pair<std::vector<Task>, xerrors::Error>
TaskClient::retrieve(const std::vector<std::string> &names) const {
    auto req = api::v1::HardwareRetrieveTaskRequest();
    req.set_rack(rack);
    req.mutable_names()->Add(names.begin(), names.end());
    auto [res, err] = task_retrieve_client->send(RETRIEVE_TASK_ENDPOINT, req);
    if (err) return {std::vector<Task>(), err};
    std::vector<Task> tasks = {res.tasks().begin(), res.tasks().end()};
    return {tasks, err};
}

std::pair<Task, xerrors::Error>
TaskClient::retrieve_by_type(const std::string &type) const {
    auto req = api::v1::HardwareRetrieveTaskRequest();
    req.set_rack(rack);
    req.add_types(type);
    auto [res, err] = task_retrieve_client->send(RETRIEVE_TASK_ENDPOINT, req);
    if (err) return {Task(), err};
    if (res.tasks_size() == 0)
        return {
            Task(),
            xerrors::Error(xerrors::NOT_FOUND, "Task matching" + type + " not found")
        };
    return {Task(res.tasks(0)), err};
}

std::pair<std::vector<Task>, xerrors::Error>
TaskClient::retrieve_by_type(const std::vector<std::string> &types) const {
    auto req = api::v1::HardwareRetrieveTaskRequest();
    req.set_rack(rack);
    req.mutable_types()->Add(types.begin(), types.end());
    auto [res, err] = task_retrieve_client->send(RETRIEVE_TASK_ENDPOINT, req);
    if (err) return {std::vector<Task>(), err};
    std::vector<Task> tasks = {res.tasks().begin(), res.tasks().end()};
    return {tasks, err};
}

xerrors::Error TaskClient::create(Task &task) const {
    auto req = api::v1::HardwareCreateTaskRequest();
    task.to_proto(req.add_tasks());
    auto [res, err] = task_create_client->send(CREATE_TASK_ENDPOINT, req);
    if (err) return err;
    if (res.tasks_size() == 0) return unexpected_missing("task");
    task.key = res.tasks().at(0).key();
    return err;
}

xerrors::Error TaskClient::del(const TaskKey key) const {
    auto req = api::v1::HardwareDeleteTaskRequest();
    req.add_keys(key);
    auto [res, err] = task_delete_client->send(DELETE_TASK_ENDPOINT, req);
    return err;
}

std::pair<std::vector<Task>, xerrors::Error> TaskClient::list() const {
    auto req = api::v1::HardwareRetrieveTaskRequest();
    req.set_rack(rack);
    auto [res, err] = task_retrieve_client->send(RETRIEVE_TASK_ENDPOINT, req);
    if (err) return {std::vector<Task>(), err};
    std::vector<Task> tasks = {res.tasks().begin(), res.tasks().end()};
    return {tasks, err};
}

const std::string RETRIEVE_DEVICE_ENDPOINT = "/hardware/device/retrieve";
const std::string CREATE_DEVICE_ENDPOINT = "/hardware/device/create";
const std::string DELETE_DEVICE_ENDPOINT = "/hardware/device/delete";

std::pair<Device, xerrors::Error>
HardwareClient::retrieve_device(const std::string &key) const {
    auto req = api::v1::HardwareRetrieveDeviceRequest();
    req.add_keys(key);
    auto [res, err] = device_retrieve_client->send(RETRIEVE_DEVICE_ENDPOINT, req);
    if (err) return {Device(), err};
    if (res.devices_size() == 0) {
        return {
            Device(),
            xerrors::Error(xerrors::NOT_FOUND, "Device matching" + key + " not found")
        };
    }
    return {Device(res.devices(0)), err};
}

std::pair<std::vector<Device>, xerrors::Error> HardwareClient::retrieve_devices(
    const std::vector<std::string> &keys,
    bool ignore_not_found
) const {
    auto req = api::v1::HardwareRetrieveDeviceRequest();
    req.mutable_keys()->Add(keys.begin(), keys.end());
    req.set_ignore_not_found(ignore_not_found);
    auto [res, err] = device_retrieve_client->send(RETRIEVE_DEVICE_ENDPOINT, req);
    std::vector<Device> devices = {res.devices().begin(), res.devices().end()};
    return {devices, err};
}

std::pair<std::vector<Device>, xerrors::Error>
HardwareClient::retrieve_devices(HardwareDeviceRetrieveRequest &req) const {
    auto api_req = api::v1::HardwareRetrieveDeviceRequest();
    req.to_proto(api_req);
    auto [res, err] = device_retrieve_client->send(RETRIEVE_DEVICE_ENDPOINT, api_req);
    if (err) return {std::vector<Device>(), err};
    std::vector<Device> devices = {res.devices().begin(), res.devices().end()};
    return {devices, err};
}

xerrors::Error HardwareClient::create_device(Device &device) const {
    auto req = api::v1::HardwareCreateDeviceRequest();
    device.to_proto(req.add_devices());
    auto [res, err] = device_create_client->send(CREATE_DEVICE_ENDPOINT, req);
    if (err) return err;
    if (res.devices_size() == 0) return unexpected_missing("device");
    device.key = res.devices().at(0).key();
    return err;
}

xerrors::Error HardwareClient::create_devices(const std::vector<Device> &devs) const {
    auto req = api::v1::HardwareCreateDeviceRequest();
    req.mutable_devices()->Reserve(static_cast<int>(devs.size()));
    for (auto &device: devs)
        device.to_proto(req.add_devices());
    auto [res, err] = device_create_client->send(CREATE_DEVICE_ENDPOINT, req);
    return err;
}

xerrors::Error HardwareClient::delete_device(const std::string &key) const {
    auto req = api::v1::HardwareDeleteDeviceRequest();
    req.add_keys(key);
    auto [res, err] = device_delete_client->send(DELETE_DEVICE_ENDPOINT, req);
    return err;
}

xerrors::Error
HardwareClient::delete_devices(const std::vector<std::string> &keys) const {
    auto req = api::v1::HardwareDeleteDeviceRequest();
    req.mutable_keys()->Add(keys.begin(), keys.end());
    auto [res, err] = device_delete_client->send(DELETE_DEVICE_ENDPOINT, req);
    return err;
}

Device::Device(const api::v1::Device &device):
    key(device.key()),
    name(device.name()),
    rack(device.rack()),
    location(device.location()),
    make(device.make()),
    model(device.model()),
    properties(device.properties()),
    configured(device.configured()) {}

Device::Device(
    std::string key,
    std::string name,
    RackKey rack,
    std::string location,
    std::string make,
    std::string model,
    std::string properties
):
    key(std::move(key)),
    name(std::move(name)),
    rack(rack),
    location(std::move(location)),
    make(std::move(make)),
    model(std::move(model)),
    properties(std::move(properties)) {}

void Device::to_proto(api::v1::Device *device) const {
    device->set_key(key);
    device->set_name(name);
    device->set_rack(rack);
    device->set_location(location);
    device->set_make(make);
    device->set_model(model);
    device->set_properties(properties);
    device->set_configured(configured);
}
}

File: core/pkg/api/hardware.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package api

import (
	"context"
	"go/types"

	"github.com/synnaxlabs/synnax/pkg/distribution/ontology"
	"github.com/synnaxlabs/synnax/pkg/service/access"
	"github.com/synnaxlabs/synnax/pkg/service/hardware"
	"github.com/synnaxlabs/synnax/pkg/service/hardware/device"
	"github.com/synnaxlabs/synnax/pkg/service/hardware/rack"
	"github.com/synnaxlabs/synnax/pkg/service/hardware/task"
	"github.com/synnaxlabs/x/errors"
	"github.com/synnaxlabs/x/gorp"
	"github.com/synnaxlabs/x/query"
	"github.com/synnaxlabs/x/validate"
)

type HardwareService struct {
	dbProvider
	accessProvider
	internal *hardware.Service
}

func NewHardwareService(p Provider) *HardwareService {
	return &HardwareService{
		dbProvider:     p.db,
		internal:       p.Service.Hardware,
		accessProvider: p.access,
	}
}

type (
	Rack   = rack.Rack
	Task   = task.Task
	Device = device.Device
)

type (
	HardwareCreateRackRequest struct {
		Racks []rack.Rack `json:"racks" msgpack:"racks"`
	}
	HardwareCreateRackResponse struct {
		Racks []rack.Rack `json:"racks" msgpack:"racks"`
	}
)

func (svc *HardwareService) CreateRack(ctx context.Context, req HardwareCreateRackRequest) (res HardwareCreateRackResponse, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Create,
		Objects: rack.OntologyIDsFromRacks(req.Racks),
	}); err != nil {
		return res, err
	}
	if err := svc.WithTx(ctx, func(tx gorp.Tx) error {
		w := svc.internal.Rack.NewWriter(tx)
		for i, r := range req.Racks {
			if err := w.Create(ctx, &r); err != nil {
				return err
			}
			req.Racks[i] = r
		}
		res.Racks = req.Racks
		return nil
	}); err != nil {
		return HardwareCreateRackResponse{}, err
	}
	return res, nil
}

type (
	HardwareRetrieveRackRequest struct {
		Keys          []rack.Key `json:"keys" msgpack:"keys"`
		Names         []string   `json:"names" msgpack:"names"`
		SearchTerm    string     `json:"search_term" msgpack:"search_term"`
		Embedded      *bool      `json:"embedded" msgpack:"embedded"`
		HostIsNode    *bool      `json:"host_is_node" msgpack:"host_is_node"`
		Limit         int        `json:"limit" msgpack:"limit"`
		Offset        int        `json:"offset" msgpack:"offset"`
		IncludeStatus bool       `json:"include_status" msgpack:"include_status"`
	}
	HardwareRetrieveRackResponse struct {
		Racks []rack.Rack `json:"racks" msgpack:"racks"`
	}
)

func (svc *HardwareService) RetrieveRack(ctx context.Context, req HardwareRetrieveRackRequest) (res HardwareRetrieveRackResponse, _ error) {
	var (
		hasSearch = len(req.SearchTerm) > 0
		hasKeys   = len(req.Keys) > 0
		hasNames  = len(req.Names) > 0
		hasLimit  = req.Limit > 0
		hasOffset = req.Offset > 0
	)
	resRacks := make([]rack.Rack, 0, len(req.Keys)+len(req.Names))
	q := svc.internal.Rack.NewRetrieve()
	if hasKeys {
		q = q.WhereKeys(req.Keys...)
	}
	if hasNames {
		q = q.WhereNames(req.Names...)
	}
	if hasSearch {
		q = q.Search(req.SearchTerm)
	}
	if hasLimit {
		q = q.Limit(req.Limit)
	}
	if hasOffset {
		q = q.Offset(req.Offset)
	}
	if req.Embedded != nil {
		q = q.WhereEmbedded(*req.Embedded)
	}
	if req.HostIsNode != nil {
		q = q.WhereNodeIsHost()
	}
	if err := q.Entries(&resRacks).Exec(ctx, nil); err != nil {
		return res, err
	}

	if req.IncludeStatus {
		for i := range resRacks {
			if s, ok := svc.internal.State.GetRack(ctx, resRacks[i].Key); ok {
				resRacks[i].Status = &s.Status
			}
		}
	}

	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Retrieve,
		Objects: rack.OntologyIDsFromRacks(resRacks),
	}); err != nil {
		return res, err
	}
	res.Racks = resRacks
	return res, nil
}

type HardwareDeleteRackRequest struct {
	Keys []rack.Key `json:"keys" msgpack:"keys"`
}

func embeddedGuard(_ gorp.Context, r Rack) error {
	if !r.Embedded {
		return nil
	}
	return errors.Wrapf(validate.Error, "cannot delete embedded rack")
}

func (svc *HardwareService) DeleteRack(
	ctx context.Context,
	req HardwareDeleteRackRequest,
) (res types.Nil, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Delete,
		Objects: rack.OntologyIDs(req.Keys),
	}); err != nil {
		return res, err
	}
	return res, svc.WithTx(ctx, func(tx gorp.Tx) error {
		exists, err := svc.internal.Device.NewRetrieve().WhereRacks(req.Keys...).Exists(ctx, tx)
		if err != nil {
			return err
		}
		if exists {
			return errors.Wrapf(validate.Error, "cannot delete rack when devices are still attached")
		}
		exists, err = svc.internal.Task.NewRetrieve().WhereInternal(false, gorp.Required()).WhereRacks(req.Keys...).Exists(ctx, tx)
		if err != nil {
			return err
		}
		if exists {
			return errors.Wrapf(validate.Error, "cannot delete rack when tasks are still attached")
		}
		w := svc.internal.Rack.NewWriter(tx)
		for _, k := range req.Keys {
			if err = w.DeleteGuard(ctx, k, embeddedGuard); err != nil {
				return err
			}
		}
		return nil
	})
}

type (
	HardwareCreateTaskRequest struct {
		Tasks []task.Task `json:"tasks" msgpack:"tasks"`
	}
	HardwareCreateTaskResponse struct {
		Tasks []task.Task `json:"tasks" msgpack:"tasks"`
	}
)

func (svc *HardwareService) CreateTask(ctx context.Context, req HardwareCreateTaskRequest) (res HardwareCreateTaskResponse, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Create,
		Objects: task.OntologyIDsFromTasks(req.Tasks),
	}); err != nil {
		return res, err
	}
	return res, svc.WithTx(ctx, func(tx gorp.Tx) error {
		w := svc.internal.Task.NewWriter(tx)
		for i, m := range req.Tasks {
			if err := w.Create(ctx, &m); err != nil {
				return err
			}
			req.Tasks[i] = m
		}
		res.Tasks = req.Tasks
		return nil
	})
}

type (
	HardwareRetrieveTaskRequest struct {
		Rack          rack.Key   `json:"rack" msgpack:"rack"`
		Keys          []task.Key `json:"keys" msgpack:"keys"`
		Names         []string   `json:"names" msgpack:"names"`
		Types         []string   `json:"types" msgpack:"types"`
		IncludeStatus bool       `json:"include_status" msgpack:"include_status"`
		Internal      *bool      `json:"internal" msgpack:"internal"`
		Snapshot      *bool      `json:"snapshot" msgpack:"snapshot"`
		SearchTerm    string     `json:"search_term" msgpack:"search_term"`
		Limit         int        `json:"limit" msgpack:"limit"`
		Offset        int        `json:"offset" msgpack:"offset"`
	}
	HardwareRetrieveTaskResponse struct {
		Tasks []task.Task `json:"tasks" msgpack:"tasks"`
	}
)

func (svc *HardwareService) RetrieveTask(
	ctx context.Context,
	req HardwareRetrieveTaskRequest,
) (res HardwareRetrieveTaskResponse, _ error) {
	var (
		hasSearch = len(req.SearchTerm) > 0
		hasKeys   = len(req.Keys) > 0
		hasNames  = len(req.Names) > 0
		hasTypes  = len(req.Types) > 0
		hasLimit  = req.Limit > 0
		hasOffset = req.Offset > 0
	)
	q := svc.internal.Task.NewRetrieve()
	if req.Internal != nil {
		q = q.WhereInternal(*req.Internal, gorp.Required())
	}
	if req.Snapshot != nil {
		q = q.WhereSnapshot(*req.Snapshot, gorp.Required())
	}
	if hasNames {
		q = q.WhereNames(req.Names...)
	}
	if hasKeys {
		q = q.WhereKeys(req.Keys...)
	}
	if hasTypes {
		q = q.WhereTypes(req.Types...)
	}
	if hasSearch {
		q = q.Search(req.SearchTerm)
	}
	if hasLimit {
		q = q.Limit(req.Limit)
	}
	if hasOffset {
		q = q.Offset(req.Offset)
	}
	if !req.Rack.IsZero() {
		q = q.WhereRacks(req.Rack)
	}
	err := q.Entries(&res.Tasks).Exec(ctx, nil)
	if err != nil {
		return res, err
	}
	if req.IncludeStatus {
		for i := range res.Tasks {
			s, ok := svc.internal.State.GetTask(ctx, res.Tasks[i].Key)
			if ok {
				res.Tasks[i].Status = &s
			}
		}
	}
	if err = svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Retrieve,
		Objects: task.OntologyIDsFromTasks(res.Tasks),
	}); err != nil {
		return HardwareRetrieveTaskResponse{}, err
	}
	return res, nil
}

type HardwareDeleteTaskRequest struct {
	Keys []task.Key `json:"keys" msgpack:"keys"`
}

func (svc *HardwareService) DeleteTask(
	ctx context.Context,
	req HardwareDeleteTaskRequest,
) (res types.Nil, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Delete,
		Objects: task.OntologyIDs(req.Keys),
	}); err != nil {
		return res, err
	}
	return res, svc.WithTx(ctx, func(tx gorp.Tx) error {
		w := svc.internal.Task.NewWriter(tx)
		for _, k := range req.Keys {
			if err := w.Delete(ctx, k, false); err != nil {
				return err
			}
		}
		return nil
	})
}

type (
	HardwareCopyTaskRequest struct {
		Key      task.Key `json:"key" msgpack:"key"`
		Name     string   `json:"name" msgpack:"name"`
		Snapshot bool     `json:"snapshot" msgpack:"snapshot"`
	}
	HardwareCopyTaskResponse struct {
		Task task.Task `json:"task" msgpack:"task"`
	}
)

func (svc *HardwareService) CopyTask(ctx context.Context, req HardwareCopyTaskRequest) (res HardwareCopyTaskResponse, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Retrieve,
		Objects: []ontology.ID{task.OntologyID(req.Key)},
	}); err != nil {
		return res, err
	}
	err := svc.WithTx(ctx, func(tx gorp.Tx) (err error) {
		res.Task, err = svc.internal.Task.NewWriter(tx).Copy(
			ctx,
			req.Key,
			req.Name,
			req.Snapshot,
		)
		return err
	})
	if err != nil {
		return HardwareCopyTaskResponse{}, err
	}
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Create,
		Objects: []ontology.ID{task.OntologyID(res.Task.Key)},
	}); err != nil {
		return HardwareCopyTaskResponse{}, err
	}
	return res, nil
}

type HardwareCreateDeviceRequest struct {
	Devices []device.Device `json:"devices" msgpack:"devices"`
}

type HardwareCreateDeviceResponse struct {
	Devices []device.Device `json:"devices" msgpack:"devices"`
}

func (svc *HardwareService) CreateDevice(ctx context.Context, req HardwareCreateDeviceRequest) (res HardwareCreateDeviceResponse, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Create,
		Objects: device.OntologyIDsFromDevices(req.Devices),
	}); err != nil {
		return res, err
	}
	return res, svc.WithTx(ctx, func(tx gorp.Tx) error {
		w := svc.internal.Device.NewWriter(tx)
		for _, d := range req.Devices {
			if err := w.Create(ctx, d); err != nil {
				return err
			}
		}
		res.Devices = req.Devices
		return nil
	})
}

type HardwareRetrieveDeviceRequest struct {
	Keys           []string   `json:"keys" msgpack:"keys"`
	Names          []string   `json:"names" msgpack:"names"`
	Makes          []string   `json:"makes" msgpack:"makes"`
	Models         []string   `json:"models" msgpack:"models"`
	Locations      []string   `json:"locations" msgpack:"locations"`
	Racks          []rack.Key `json:"racks" msgpack:"racks"`
	SearchTerm     string     `json:"search_term" msgpack:"search_term"`
	Limit          int        `json:"limit" msgpack:"limit"`
	Offset         int        `json:"offset" msgpack:"offset"`
	IgnoreNotFound bool       `json:"ignore_not_found" msgpack:"ignore_not_found"`
	IncludeStatus  bool       `json:"include_status" msgpack:"include_status"`
}

type HardwareRetrieveDeviceResponse struct {
	Devices []device.Device `json:"devices" msgpack:"devices"`
}

func (svc *HardwareService) RetrieveDevice(ctx context.Context, req HardwareRetrieveDeviceRequest) (res HardwareRetrieveDeviceResponse, _ error) {
	var (
		hasSearch    = len(req.SearchTerm) > 0
		hasKeys      = len(req.Keys) > 0
		hasNames     = len(req.Names) > 0
		hasMakes     = len(req.Makes) > 0
		hasLimit     = req.Limit > 0
		hasOffset    = req.Offset > 0
		hasLocations = len(req.Locations) > 0
		hasModels    = len(req.Models) > 0
		hasRacks     = len(req.Racks) > 0
	)
	q := svc.internal.Device.NewRetrieve()
	if hasKeys {
		q = q.WhereKeys(req.Keys...)
	}
	if hasSearch {
		q = q.Search(req.SearchTerm)
	}
	if hasNames {
		q = q.WhereNames(req.Names...)
	}
	if hasLimit {
		q = q.Limit(req.Limit)
	}
	if hasOffset {
		q = q.Offset(req.Offset)
	}
	if hasMakes {
		q = q.WhereMakes(req.Makes...)
	}
	if hasLocations {
		q = q.WhereLocations(req.Locations...)
	}
	if hasModels {
		q = q.WhereModels(req.Models...)
	}
	if hasRacks {
		q = q.WhereRacks(req.Racks...)
	}
	retErr := q.Entries(&res.Devices).Exec(ctx, nil)
	if req.IncludeStatus {
		for i := range res.Devices {
			if s, ok := svc.internal.State.GetDevice(ctx, res.Devices[i].Key); ok {
				res.Devices[i].Status = &s
			}
		}
	}
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Retrieve,
		Objects: device.OntologyIDsFromDevices(res.Devices),
	}); err != nil {
		return HardwareRetrieveDeviceResponse{}, err
	}
	if retErr != nil && req.IgnoreNotFound {
		retErr = errors.Skip(retErr, query.NotFound)
	}
	return res, retErr
}

type HardwareDeleteDeviceRequest struct {
	Keys []string `json:"keys" msgpack:"keys"`
}

func (svc *HardwareService) DeleteDevice(ctx context.Context, req HardwareDeleteDeviceRequest) (res types.Nil, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Delete,
		Objects: device.OntologyIDs(req.Keys),
	}); err != nil {
		return res, err
	}
	return res, svc.WithTx(ctx, func(tx gorp.Tx) error {
		w := svc.internal.Device.NewWriter(tx)
		for _, k := range req.Keys {
			if err := w.Delete(ctx, k); err != nil {
				return err
			}
		}
		return nil
	})
}
